lighthouse-core/test/lib/tracehouse/main-thread-tasks-test.js

```diff
@@ -468,25 +468,74 @@ describe('Main Thread Tasks', () => {
     ]);
   });
 
+  it('should handle child events that extend >1ms beyond parent event because missing E', () => {
+    /*
+    An artistic rendering of the below trace:
+    ████████████████TaskA██████████████████
+            █████████TaskB██████████████████
+    */
+    const traceEvents = [
+      ...boilerplateTrace,
+      {ph: 'B', name: 'TaskA', pid, tid, ts: baseTs, args},
+      {ph: 'B', name: 'TaskB', pid, tid, ts: baseTs + 25e3, args},
+      {ph: 'E', name: 'TaskA', pid, tid, ts: baseTs + 100e3, args},
+      {ph: 'I', name: 'MarkerToPushOutTraceEnd', pid, tid, ts: baseTs + 110e3, args},
+    ];
+
+    traceEvents.forEach(evt => Object.assign(evt, {cat: 'devtools.timeline'}));
+
+    const tasks = run({traceEvents});
+    const [taskA, taskB] = tasks;
+    expect(tasks).toEqual([
+      {
+        parent: undefined,
+        attributableURLs: [],
+
+        children: [taskB],
+        event: traceEvents.find(event => event.name === 'TaskA'),
+        startTime: 0,
+        endTime: 100,
+        duration: 100,
+        selfTime: 25,
+        group: taskGroups.other,
+      },
+      {
+        parent: taskA,
+        attributableURLs: [],
+
+        children: [],
+        event: traceEvents.find(event => event.name === 'TaskB' && event.ph === 'B'),
+        startTime: 25,
+        endTime: 100,
+        duration: 75,
+        selfTime: 75,
+        group: taskGroups.other,
+      },
+    ]);
+  });
+
   const invalidEventSets = [
     [
       // TaskA overlaps with TaskB, X first
       {ph: 'X', name: 'TaskA', pid, tid, ts: baseTs, dur: 100e3, args},
       {ph: 'B', name: 'TaskB', pid, tid, ts: baseTs + 5e3, args},
       {ph: 'E', name: 'TaskB', pid, tid, ts: baseTs + 115e3, args},
+      {ph: 'I', name: 'MarkerToPushOutTraceEnd', pid, tid, ts: baseTs + 200e3, args},
```

Comment:
some of these seem like they won't affect anything, is it just general hygiene so they're all similar?

Reply:
they actually do have a noticeable effect though because if the `E` is the traceEnd then we fall into a different error recovery mode that turns these into valid events which is not what we want to test

Reply:
added a comment to this effect

Reply:
> they actually do have a noticeable effect though because if the `E` is the traceEnd then we fall into a different error recovery mode that turns these into valid events which is not what we want to test

oh right, because then it would be assumed TaskB didn't have an end event (even though it did) and one was created for it (even though one wasn't), so then it would get trimmed to the length of TaskA.

Doesn't that seem like a problem with using `traceEndTs` as the signal we were missing an end event? It seems like this case (before the `MarkerToPushOutTraceEnd` event is added) shouldn't be treated that way (even if it is very unlikely to occur).

Are there downsides to using `Infinity` for the `ts` in https://github.com/GoogleChrome/lighthouse/blob/61f3ee448c9c38f69d4af24439217d8affad63b5/lighthouse-core/lib/tracehouse/main-thread-tasks.js#L187

?

Reply:
> Doesn't that seem like a problem with using traceEndTs as the signal we were missing an end event? It seems like this case (before the MarkerToPushOutTraceEnd event is added) shouldn't be treated that way (even if it is very unlikely to occur).

I disagree. In the case where we have just one of the E events and no other events following it at all, the most likely explanation to me is that it just got cutoff emitting the other `E` event for the exact same time, i.e. the exact case our workaround is targeting. 

The fact that other events follow it is what makes it a completely invalid situation.

Reply:
> In the case where we have just one of the E events and no other events following it at all, the most likely explanation to me is that it just got cutoff emitting the other `E` event for the exact same time, i.e. the exact case our workaround is targeting

Hmm, I don't understand. In the case of
```
{ph: 'X', name: 'TaskA', pid, tid, ts: baseTs, dur: 100e3, args},
{ph: 'B', name: 'TaskB', pid, tid, ts: baseTs + 5e3, args},
{ph: 'E', name: 'TaskB', pid, tid, ts: baseTs + 115e3, args},
```
clearly there was an end event emitted...it seems better to check if other invalid heuristics apply than to just chop off TaskB at `baseTs + 100e3`.

Reply:
Oh, oh sorry! Yeah this specific case sounds like it should work without it. 👍

Reply:
alright `Infinity` has it's own set of problems which is why I avoided it to begin with, so introduced an `unbounded` boolean to deal with this. LMK if you think it's worth the tradeoff in the latest commit

Reply:
> alright `Infinity` has it's own set of problems which is why I avoided it to begin with, so introduced an `unbounded` boolean to deal with this. LMK if you think it's worth the tradeoff in the latest commit

I figured there must be dragons in Infinity, but it was worth a look :) `unbounded` isn't the most elegant looking, but IMO it's worth it.